### PR TITLE
Make LWS_SEND_BUFFER_PRE_PADDING preprocessor if-friendly

### PR DESCRIPTION
**Pull request by alukichev**
Commit 173e9c4e made LWS_SEND_BUFFER_SIZE a multiple of a certain
value returned by _LWS_PAD_SIZE macro. This macro expanded to
"sizeof(void *)" on non-x86_64 architectures, which made it
unsuitable to use LWS_SEND_BUFFER_SIZE in preprocessor #if
expressions in the library user code.

This patch preserves the padding logic since commit 173e9c4e but
makes it more preprocessor-friendly for applications using
libwebsockets by setting _LWS_PAD_SIZE to the size of "void *"
determined by cmake when libwebsockets is configured for the
target platform.

Signed-off-by: Alexander Lukichev <alexander.lukichev@gmail.com>

### CODE REVIEW

File: lws_config.h.in [hunk 1]
```diff
@@ -61,3 +61,5 @@
 
 /* use SHA1() not internal libwebsockets_SHA1 */
 #cmakedefine LWS_SHA1_USE_OPENSSL_NAME
+
+${LWS_SIZEOFPTR_CODE}
```
Comment (JoakimSoderberg) [inline lws_config.h.in]: Use `#cmakedefine LWS_SIZEOFPTR_CODE`

Comment (JoakimSoderberg) [inline lws_config.h.in]: Hrm, I mean how did you get this to work at all? Because this code is broken.... You'll have a header with a random `8` or whatever in it, no define or nothing...

Comment (alukichev) [inline lws_config.h.in]: @JoakimSoderberg wrote:

> Use #cmakedefine LWS_SIZEOFPTR_CODE

Using this doesn't make much sense. It expands to "#define LWS_SIZEOFPTR_CODE" if cmake could determine the size of "void *" (almost always). As a result, _LWS_PAD_SIZE becomes an empty constant and evaluates to 0 in _LWS_PAD() macro.

From [cmake docs](https://cmake.org/cmake/help/v3.0/command/configure_file.html):
> Input file lines of the form “#cmakedefine VAR ...” will be replaced with either “#define VAR ...” or /* #undef VAR */ depending on whether VAR is set in CMake to any value not considered a false constant by the if() command.

On the other hand, [CHECK_TYPE_SIZE()](https://cmake.org/cmake/help/v3.0/module/CheckTypeSize.html) allows one to generate preprocessor code that expands to a number, even for an OS X multi-architecture binary. See the documentation for this macro.

Comment (JoakimSoderberg) [inline lws_config.h.in]: ah ok I missed the `_CODE` part... I did not know that part. I have only ever seen people define it like `#cmakedefine LWS_SIZEOFPTR_CODE ${LWS_SIZEOFPTR_CODE}`

Comment (alukichev) [inline lws_config.h.in]: `#cmakedefine LWS_SIZEOFPTR ${LWS_SIZEOFPTR}` would work here, in most cases. Originally, I thought to put it there too but in the end I decided to use just `${LWS_SIZEOFPTR_CODE}` in the lws_config.h.in for the following reasons.

Variable `LWS_SIZEOFPTR` in the CMakeLists.txt gets its value from `CHECK_TYPE_SIZE()` macro, which can [return](https://cmake.org/cmake/help/v3.0/module/CheckTypeSize.html):
* `<size>` - an integer literal if the type size could be determined - for "void *" this would be always true except for certain cases on OS X;
* `"0"` - **a string literal** when building libwebsockets as a universal binary-ready library - cmake will do that on OS X if you use it as e.g., `cmake "-DCMAKE_OSX_ARCHITECTURES=x86_64;i386" <path>`;
* `""` - an empty string - should not be normally possible for "void *".

Thus, in most cases, simply using `#cmakedefine VAR ${VAR}` would result in a `#define VAR <size>` in the lws_config.h file, which would make `_LWS_PAD()` macro in libwebsockets.h work correctly. In case of a multi-architecture build on OS X, though, the build would break completely because there would be `#define VAR "0"` in lws_config.h, which would be painful to filter out with preprocessor directives (if at all possible). It could be analyzed in CMakeLists.txt and substituted with 0 but it would disable the whole feature introduced by commit 173e9c4 for the universal binary use case. So, using `#cmakedefine` would make some users of libwebsockets unhappy if they wanted to build it as a "universal binary"-ready library for their hypothetical multi-architecture OS X application.

The same `CHECK_TYPE_SIZE(type VAR)` expression has on the output actually 3 variables:
* `HAVE_VAR` - if the type exists - true for "void *";
* `VAR` - previously discussed;
* `VAR_CODE` - a set of preprocessor directives to define `VAR` by preprocessor to the correct value, or `/* #undef VAR*/` if the type does not exist (not true in case of "void *").

This last cmake variable expands to a simple `#define VAR <size>` in most mentioned cases, just like `#cmakedefine` does. In case of OS X multi-architecture build, though, it [expands](https://cmake.org/gitweb?p=cmake.git;a=blob;f=Modules/CheckTypeSize.cmake;h=5938d6c9a720dd67124affda8bf6707769d2fb51;hb=HEAD#l256) to a proper set of `#if defined(<arch1>) ... #elif defined(<arch2>) ... #else #error ... #endif` directives, which will work correctly when building the same library for multiple architectures. Thus, it does not introduce a limitation that `#cmakedefine` approach does. *More importantly, since this goes to lws_config.h, it will be included by user code, and will generate correct pointer size for each architecture if that user code happens to be a multi-architecture binary*. Using `#cmakedefine` would put a single number there (if "0" literal is substituted with something sensible).

Using `#cmakedefine VAR ${VAR}` in lws_config.h.in to define the size of "void *" is completely identical to simply using `#define VAR ${VAR}` since that type is a built-in type, and VAR would always be non-empty. `#cmakedefine` is more suited to `HAVE_<feature>` use cases where `<feature>` is much more probable to be absent on a platform.

To summarize, using `${VAR_CODE}` instead of `#cmakedefine VAR ${VAR}` is a compact way of achieving the same goal using the same means but is not redundant and is more portable.

If you still would rather prefer to change that line to `#cmakedefine LWS_SIZEOFPTR ${LWS_SIZEOFPTR}`, I'll do that and will supply a new version of this patch. Thanks.

Comment (JoakimSoderberg) [inline lws_config.h.in]: Thanks for the detailed explanation. This is all good. I just hadn't been aware of the `${VAR_CODE}` stuff. This is really useful to know about :)

Your patch is fine as it is!

**Comment by JoakimSoderberg** [comment]
Merged thanks
